File: Mridul/PS26167/mvp/satquery/evidence.py
"""The evidence layer.

Every specialist writes into one normalised record. That record is the contract
between the machine-learning half of the system and the product half, and it is
what makes ADR-007 enforceable rather than aspirational: the answer generator
is handed `Evidence`, never pixels, so it cannot invent a number it was not
given.

Three things happen here that matter:

1. **Pixel coordinates become Earth coordinates.** A box at pixels
   (120, 50, 220, 170) is not useful. The same box at 23.41 N 85.32 E,
   exported as GeoJSON, opens in QGIS.

2. **Confidence is gated.** Below threshold the system abstains rather than
   answering. Abstention is a feature, and its precision is measured.

3. **Conflicts are recorded, not resolved silently.** When optical and SAR
   disagree, the disagreement is part of the evidence and the confidence drops.
"""

from __future__ import annotations

import json
import math
from dataclasses import dataclass, field, asdict
from typing import Any, Literal

import numpy as np

from .raster import GeoTransform, Raster

Modality = Literal["optical", "sar", "fused", "temporal", "derived"]


# --------------------------------------------------------------------------- #
# geometry
# --------------------------------------------------------------------------- #

@dataclass
class GeoBox:
    """A rectangle, carried in both pixel and world coordinates."""

    x0: int
    y0: int
    x1: int
    y1: int
    lon0: float = 0.0
    lat0: float = 0.0
    lon1: float = 0.0
    lat1: float = 0.0
    area_px: int = 0
    area_ha: float = 0.0

    @classmethod
    def from_pixels(cls, bbox: list[int], transform: GeoTransform,
                    area_px: int = 0) -> "GeoBox":
        x0, y0, x1, y1 = bbox
        lon0, lat0 = transform.pixel_to_world(x0, y0)
        lon1, lat1 = transform.pixel_to_world(x1 + 1, y1 + 1)
        gsd = transform.ground_sample_distance
        area = area_px or ((x1 - x0 + 1) * (y1 - y0 + 1))
        return cls(x0, y0, x1, y1,
                   round(lon0, 6), round(lat0, 6), round(lon1, 6), round(lat1, 6),
                   int(area), round(area * gsd * gsd / 10_000.0, 3))

    def centre(self) -> tuple[float, float]:
        return (self.lat0 + self.lat1) / 2, (self.lon0 + self.lon1) / 2

    def to_geojson(self, props: dict | None = None) -> dict:
        return {
            "type": "Feature",
            "geometry": {"type": "Polygon", "coordinates": [[
                [self.lon0, self.lat0], [self.lon1, self.lat0],
                [self.lon1, self.lat1], [self.lon0, self.lat1],
                [self.lon0, self.lat0]]]},
            "properties": {"area_ha": self.area_ha, "area_px": self.area_px,
                           **(props or {})},
        }


# --------------------------------------------------------------------------- #
# evidence
# --------------------------------------------------------------------------- #

@dataclass
class Evidence:
    """One claim, its support, and where it came from."""

    claim: str
    value: Any = None
    unit: str = ""
    confidence: float = 0.0
    modality: Modality = "derived"
    source_model: str = ""
    source_version: str = ""
    boxes: list[GeoBox] = field(default_factory=list)
    mask_area_ha: float = 0.0
    supporting: list[str] = field(default_factory=list)
    conflicts: list[str] = field(default_factory=list)
    method: str = ""

    def to_dict(self) -> dict:
        d = asdict(self)
        d["boxes"] = [asdict(b) for b in self.boxes]
        d["confidence"] = round(float(self.confidence), 3)
        return d


@dataclass
class EvidenceSet:
    """Everything gathered for one query, plus the gate applied to it."""

    items: list[Evidence] = field(default_factory=list)
    threshold: float = 0.45
    crs: str = "EPSG:4326"

    def add(self, ev: Evidence) -> "EvidenceSet":
        self.items.append(ev)
        return self

    # -- gating ------------------------------------------------------------ #
    @property
    def passing(self) -> list[Evidence]:
        return [e for e in self.items if e.confidence >= self.threshold]

    @property
    def abstain(self) -> bool:
        """True when nothing cleared the gate — the system declines to answer."""
        return len(self.items) > 0 and len(self.passing) == 0

    @property
    def confidence(self) -> float:
        """Area-weighted mean over passing evidence, penalised for conflict.

        Weighted by area because a claim resting on one 12-pixel blob should
        not carry the same weight as one resting on a hectare, and penalised
        for conflict because two sensors disagreeing is a real reason to be
        less sure.
        """
        p = self.passing
        if not p:
            return 0.0
        weights = [max(e.mask_area_ha, 0.01) if e.mask_area_ha else 1.0 for e in p]
        total = sum(weights)
        base = sum(e.confidence * w for e, w in zip(p, weights)) / max(total, 1e-9)
        conflicts = sum(len(e.conflicts) for e in p)
        return round(float(max(0.0, base - 0.08 * conflicts)), 3)

    # -- export ------------------------------------------------------------ #
    def geojson(self) -> dict:
        feats = []
        for e in self.passing:
            for b in e.boxes:
                feats.append(b.to_geojson({
                    "claim": e.claim, "confidence": round(e.confidence, 3),
                    "modality": e.modality, "model": e.source_model,
                }))
        return {"type": "FeatureCollection",
                "crs": {"type": "name", "properties": {"name": self.crs}},
                "features": feats}

    def to_dict(self) -> dict:
        return {
            "threshold": self.threshold,
            "confidence": self.confidence,
            "abstain": self.abstain,
            "count": len(self.items),
            "passing": len(self.passing),
            "items": [e.to_dict() for e in self.items],
        }

    def dumps(self) -> str:
        return json.dumps(self.to_dict(), indent=2)


# --------------------------------------------------------------------------- #
# helpers
# --------------------------------------------------------------------------- #

def boxes_from_props(props: list[dict], transform: GeoTransform,
                     limit: int = 24) -> list[GeoBox]:
    return [GeoBox.from_pixels(p["bbox"], transform, p["area_px"])
            for p in props[:limit]]


def mask_area_ha(mask: np.ndarray, transform: GeoTransform) -> float:
    gsd = transform.ground_sample_distance
    return round(float(mask.sum()) * gsd * gsd / 10_000.0, 3)


def confidence_from_separation(values: np.ndarray, threshold: float) -> float:
    """Confidence derived from how cleanly a threshold splits the data.

    A threshold sitting in a deep valley between two well-separated modes is
    trustworthy. One sitting inside a single broad blob is not, and the answer
    that rests on it should say so. This is what stops every detection being
    reported at a flat, meaningless 0.9.
    """
    v = values[np.isfinite(values)]
    if v.size < 16:
        return 0.35
    below, above = v[v < threshold], v[v >= threshold]
    if below.size < 4 or above.size < 4:
        return 0.40
    m0, m1 = float(below.mean()), float(above.mean())
    s0, s2 = float(below.std()), float(above.std())
    pooled = math.sqrt((s0 * s0 + s2 * s2) / 2) or 1e-6
    d = abs(m1 - m0) / pooled                       # Cohen's d
    return round(float(min(0.97, 0.42 + 0.16 * d)), 3)

File: 167/web/src/engine/evidence.ts
/**
 * The evidence layer — a port of `satquery/evidence.py`.
 *
 * Pixel coordinates become Earth coordinates; confidence is gated; conflicts
 * are recorded rather than silently resolved.
 */

import type { EvidenceItem, GeoBox, Modality } from '@/lib/contract'
import { pixelAreaM2, pixelToLonLat, type Raster } from './raster'
import type { Region } from './cv'

const r6 = (v: number) => Math.round(v * 1e6) / 1e6
const r3 = (v: number) => Math.round(v * 1e3) / 1e3

export function geoBox(r: Raster, bbox: [number, number, number, number], areaPx = 0): GeoBox {
  const [x0, y0, x1, y1] = bbox
  const a = pixelToLonLat(r, x0, y0)
  const b = pixelToLonLat(r, x1 + 1, y1 + 1)
  const area = areaPx || (x1 - x0 + 1) * (y1 - y0 + 1)
  return {
    x0, y0, x1, y1,
    lon0: a ? r6(a[0]) : null, lat0: a ? r6(a[1]) : null,
    lon1: b ? r6(b[0]) : null, lat1: b ? r6(b[1]) : null,
    area_px: area,
    area_ha: r3(area * pixelAreaM2(r, (x0 + x1) / 2, (y0 + y1) / 2) / 10_000),
  }
}

export const boxesFromProps = (props: Region[], r: Raster, limit = 24) =>
  props.slice(0, limit).map((p) => geoBox(r, p.bbox, p.area_px))

/** Measured at the scene centre; the backend uses the mask's centroid — within a
 * scene-sized image the two differ by well under 1 %. */
export function maskAreaHa(count: number, r: Raster): number {
  return r3(count * pixelAreaM2(r, r.width / 2, r.height / 2) / 10_000)
}

/**
 * Confidence from how cleanly a threshold splits the data (Cohen's d between
 * the two sides). Stops every detection being reported at a flat 0.9.
 */
export function confidenceFromSeparation(values: Float32Array, thr: number): number {
  let n0 = 0, n1 = 0, s0 = 0, s1 = 0, q0 = 0, q1 = 0, total = 0
  for (let i = 0; i < values.length; i++) {
    const v = values[i]
    if (!Number.isFinite(v)) continue
    total++
    if (v < thr) { n0++; s0 += v; q0 += v * v } else { n1++; s1 += v; q1 += v * v }
  }
  if (total < 16) return 0.35
  if (n0 < 4 || n1 < 4) return 0.40
  const m0 = s0 / n0, m1 = s1 / n1
  const sd0 = Math.sqrt(Math.max(q0 / n0 - m0 * m0, 0))
  const sd1 = Math.sqrt(Math.max(q1 / n1 - m1 * m1, 0))
  const pooled = Math.sqrt((sd0 * sd0 + sd1 * sd1) / 2) || 1e-6
  const d = Math.abs(m1 - m0) / pooled
  return r3(Math.min(0.97, 0.42 + 0.16 * d))
}

export function ev(source: { model: string; version: string }, fields: Partial<EvidenceItem> & { claim: string }): EvidenceItem {
  return {
    value: null, unit: '', confidence: 0, modality: 'derived' as Modality,
    boxes: [], mask_area_ha: 0, supporting: [], conflicts: [], method: '',
    ...fields,
    source_model: source.model,
    source_version: source.version,
  }
}

export class EvidenceSet {
  items: EvidenceItem[] = []
  threshold: number
  crs = 'EPSG:4326'

  constructor(threshold = 0.45) { this.threshold = threshold }

  add(e: EvidenceItem) { this.items.push(e); return this }

  get passing() { return this.items.filter((e) => e.confidence >= this.threshold) }

  /** True when nothing cleared the gate — the system declines to answer. */
  get abstain() { return this.items.length > 0 && this.passing.length === 0 }

  /** Area-weighted mean over passing evidence, minus 0.08 per recorded conflict. */
  get confidence() {
    const p = this.passing
    if (!p.length) return 0
    const w = p.map((e) => (e.mask_area_ha ? Math.max(e.mask_area_ha, 0.01) : 1))
    const total = w.reduce((a, b) => a + b, 0)
    const base = p.reduce((a, e, i) => a + e.confidence * w[i], 0) / Math.max(total, 1e-9)
    const conflicts = p.reduce((a, e) => a + e.conflicts.length, 0)
    return r3(Math.max(0, base - 0.08 * conflicts))
  }

  geojson() {
    const features = []
    for (const e of this.passing) {
      for (const b of e.boxes) {
        if (b.lon0 === null || b.lat0 === null || b.lon1 === null || b.lat1 === null) continue
        features.push({
          type: 'Feature' as const,
          geometry: {
            type: 'Polygon' as const,
            coordinates: [[[b.lon0, b.lat0], [b.lon1, b.lat0], [b.lon1, b.lat1], [b.lon0, b.lat1], [b.lon0, b.lat0]]],
          },
          properties: {
            area_ha: b.area_ha, area_px: b.area_px, claim: e.claim,
            confidence: r3(e.confidence), modality: e.modality, model: e.source_model,
          },
        })
      }
    }
    return { type: 'FeatureCollection', crs: { type: 'name', properties: { name: this.crs } }, features }
  }

  toDict() {
    return {
      threshold: this.threshold,
      confidence: this.confidence,
      abstain: this.abstain,
      count: this.items.length,
      passing: this.passing.length,
      items: this.items.map((e) => ({ ...e, confidence: r3(e.confidence) })),
    }
  }
}
